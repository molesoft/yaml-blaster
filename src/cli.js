/*
 * This file is subject to the terms and conditions defined in
 * file 'LICENSE', which is part of this source code package.
 */

const {readFileSync, writeFileSync} = require('fs')
const {join} = require('path')
const yargs = require('yargs')
const chalk = require('chalk')
const boxen = require('boxen')
const Blaster = require('./blaster')
const yaml = require('js-yaml')

const args = yargs
  .usage("Usage: yaml-blaster -i <input path> -d <data path> -0 <output path>")
  .option("i", { alias: "input", describe: "Input file path", type: "string", demandOption: true })
  .option("d", { alias: "data", describe: "Data file path", type: "string", demandOption: true })
  .option("o", { alias: "out", describe: "Output file path", type: "string", demandOption: false })
  .option("p", { alias: "params", describe: "Additional parameters (ex: key1=val1,key2=val2)", type: "string", demandOption: false })
  .argv

const relDir = process.cwd()
const inputPath = args.input.startsWith('/') ? args.input : join(relDir, args.input)
const inputDir = inputPath.split('/').reduce((s,item,i) => {
  return i === inputPath.split('/').length - 1 ? s : `${s}/${item}`
}, '')
const input = readFileSync(inputPath, 'utf8')
const dataString = readFileSync(args.data, 'utf8')
let data
try {
  data = JSON.parse(dataString)
} catch(err) {
  data = yaml.safeLoad(dataString)
}

let params
if(args.params) {
  try {
    params = args.params.split(',').map(param => {
      const keyVal = param.split('=')
      return {
        key: keyVal[0],
        val: keyVal[1]
      }
    })
  } catch(err) {
    throw new Error('Please supply params in the following format: key=val,key2=val2')
  }
}
const yb = new Blaster(input, data, inputDir)
const processed = yb.process(input, data, inputDir, params)
if(args.out) {
  const outPath = args.out.startsWith('/') ? args.out : join(relDir, args.out)
  writeFileSync(outPath, processed)
} else {
  console.log(processed)
}