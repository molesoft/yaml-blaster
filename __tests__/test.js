const Blaster = require('../dist/blaster')
const {join} = require('path')
const {readFileSync, unlinkSync} = require('fs')
const yaml = require('js-yaml')
const uuid = require('uuid')

it('should handle expressions', () => {
  const inputPath = join(__dirname,'logic')
  const inputFile = join(inputPath, 't.yaml')
  const dataPath = join(inputPath, 'd.yaml')
  const data = yaml.safeLoad(readFileSync(dataPath,'utf8'))
  const input = readFileSync(inputFile, 'utf8')
  const yb = new Blaster(input, data, inputPath)
  const result = yb.process(input, data, inputPath)
  expect(result).toMatchSnapshot()
})

it('should handle complex comments and multiple vars on the same line', () => {
  const inputPath = join(__dirname,'sample2')
  const inputFile = join(inputPath, 'template.yaml')
  const dataPath = join(inputPath, 'data.yaml')
  const data = yaml.safeLoad(readFileSync(dataPath,'utf8'))
  const input = readFileSync(inputFile, 'utf8')
  const yb = new Blaster(input, data, inputPath)
  const result = yb.process(input, data, inputPath)
  expect(result).toMatchSnapshot()
})

it('should have a cli interface that works', () => {
  const argvBackup = [...process.argv]
  const args = [...process.argv]
  const tmpPath = `/tmp/${uuid.v4()}`
  args.push('-i', '__tests__/samples/template.yaml', '-d', '__tests__/samples/data.json', '-o', tmpPath, '-p', 'myKey=myVal,myKey2=myVal2')
  process.argv = args
  require('../dist/cli')
  process.argv = argvBackup
  const result = readFileSync(tmpPath, 'utf8')
  unlinkSync(tmpPath)
  expect(result).toMatchSnapshot()
})

it('should handle mutliple stages', async () => {
  const data = {
    baz: 'bazval',
    cat: 'catval',
    cidr: '0.0.0.0/0',
    proto: 'https',
    someMap: {
      baz: 'mapbaz',
      cat: 'mapcat'
    },
    instances: [
      {
        name: 'dog',
        size: 'big',
        cidr: '0.1.1.1',
        tags: {
          name: 'tagname',
          val: 'tagval'
        }
      },
      {
        name: 'cat',
        size: 'real big',
        cidr: '0.2.2.2',
        tags: {
          name: 'tagname2',
          val: 'tagval2'
        }
      }
    ]
  }
  const inputPath = join(__dirname,'samples')
  const inputFile = join(inputPath, 'template.yaml')
  const input = readFileSync(inputFile, 'utf8')
  const yb = new Blaster(input, data, inputPath)
  const result = yb.process(input, data, inputPath)
  expect(result).toMatchSnapshot()
})

it('should update var paths', async () => {
  const fragment = `
  {{name}}:
  Type: AWS::EC2::Instance
  Properties:
    Size: {{size}}
    Cidr: {{cidr}}
  Tags: {{file:subTemplates/subSubTemplates/tags.yaml}}
  Foo: {{for:tags, file:subTemplates/subSubTemplates/tags.yaml}}
  `
  const yb = new Blaster(fragment)
  const result = yb.updateVarPaths(fragment, 'instances.0')
  expect(result).toMatchSnapshot()
})

it('should flatten a complex object', async () => {
  const data = {
    foo: 'bar',
    instances: [
      { name: 'dog'  },
      { name: 'cat' }
    ],
    baz: [
      [1,2],
      [3,4],
      [{
        test: [
          4,5,{5:6}
        ]
      }]
    ]
  }
  const yb = new Blaster('')
  const result = yb.flatten(data)
  expect(result).toMatchSnapshot()
})

it('should handle additional params', async () => {
  const data = {
    foo: 'bar'
  }
  const input = `
  Test: {{myKey}}
  Test2: {{myKey2}}
  Test3: {{foo}}
  `
  const params = [
    {
      key: 'myKey',
      val: 'myVal'
    },
    {
      key: 'myKey2',
      val: 'myVal2'
    }
  ]
  const yb = new Blaster(input, data, null)
  const result = yb.process(input, data, null, params)
  expect(result).toMatchSnapshot()
})

it.skip('should handle functions', async () => {
  const data = {
    foo: 'bar'
  }
  const input = `
  Test: {{$epoch}}
  Test2: {{$uuid}}
  Test3: {{$epoch}}
  Test4: {{$uuid}}
  `
  const yb = new Blaster(input, data)
  const result = yb.process(input, data)
  console.log(result)
  // expect(result).toMatchSnapshot()
})