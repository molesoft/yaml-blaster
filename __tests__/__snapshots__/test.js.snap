// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should flatten a complex object 1`] = `
Object {
  "baz.0.0": 1,
  "baz.0.1": 2,
  "baz.1.0": 3,
  "baz.1.1": 4,
  "baz.2.0.test.0": 4,
  "baz.2.0.test.1": 5,
  "baz.2.0.test.2.5": 6,
  "foo": "bar",
  "instances.0.name": "dog",
  "instances.1.name": "cat",
}
`;

exports[`should handle additional params 1`] = `
"Test: myVal
Test2: myVal2
Test3: bar
"
`;

exports[`should handle complex comments and multiple vars on the same line 1`] = `
"AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'
Description: Setup for network resources
Resources:
  VPC:
    Type: 'AWS::EC2::VPC'
    Properties:
      CidrBlock: 10.0.0.0/16
      EnableDnsHostnames: 'true'
      EnableDnsSupport: 'true'
      InstanceTenancy: default
      Tags:
        - Key: Name
          Value:
            'Fn::Sub': '\${AWS::StackName}-uscc-vpc'
  PublicSubneta:
    Type: 'AWS::EC2::Subnet'
    Properties:
      VpcId:
        Ref: VPC
      CidrBlock: 10.0.0.0/20
      AvailabilityZone:
        'Fn::Sub': us-east-1a
      Tags:
        - Key: Name
          Value:
            'Fn::Sub': '\${AWS::StackName}-us-east-1-a-Public'
  PublicSubnetaAssociation:
    Type: 'AWS::EC2::SubnetRouteTableAssociation'
    Properties:
      SubnetId:
        Ref: PublicSubneta
      RouteTableId:
        Ref: PublicRouteTable
  InternetGateway:
    Type: 'AWS::EC2::InternetGateway'
    Properties:
      Tags:
        - Key: Name
          Value:
            'Fn::Sub': '\${AWS::StackName}-uscc-igw'
  AttachGateway:
    Type: 'AWS::EC2::VPCGatewayAttachment'
    Properties:
      VpcId:
        Ref: VPC
      InternetGatewayId:
        Ref: InternetGateway
  PublicRouteTable:
    Type: 'AWS::EC2::RouteTable'
    Properties:
      VpcId:
        Ref: VPC
      Tags:
        - Key: Name
          Value:
            'Fn::Sub': '\${AWS::StackName}-public-table'
  PublicRoute:
    Type: 'AWS::EC2::Route'
    Properties:
      RouteTableId:
        Ref: PublicRouteTable
      DestinationCidrBlock: 0.0.0.0/0
      GatewayId:
        Ref: InternetGateway
"
`;

exports[`should handle expressions 1`] = `
"Foo: barrrrr
Baz: bazzzzz
Bar:
  dog: yay
  cat: nay
Bar2:
  dog: yay
  cat: nay
Bar3: barrrrr
dog: yay
cat: nay
asdf:
  dog: yay
  cat: nay
subTest:
  answer: not meow
"
`;

exports[`should handle mutliple stages 1`] = `
"AWSTemplateFormatVersion: '2010-09-09'
dog:
  Type: 'AWS::EC2::Instance'
  Properties:
    Size: big
    Cidr: 0.1.1.1
  Tags:
    Name: tagname
    Value: tagval
cat:
  Type: 'AWS::EC2::Instance'
  Properties:
    Size: real big
    Cidr: 0.2.2.2
  Tags:
    Name: tagname2
    Value: tagval2
Resources:
  dog:
    Type: 'AWS::EC2::Instance'
    Properties:
      Size: big
      Cidr: 0.1.1.1
    Tags:
      Name: tagname
      Value: tagval
  cat:
    Type: 'AWS::EC2::Instance'
    Properties:
      Size: real big
      Cidr: 0.2.2.2
    Tags:
      Name: tagname2
      Value: tagval2
  NotificationTopic:
    Type: 'AWS::SNS::Topic'
    Properties:
      Subscription:
        - dog:
            Type: 'AWS::EC2::Instance'
            Properties:
              Size: big
              Cidr: 0.1.1.1
            Tags:
              Name: tagname
              Value: tagval
        - cat:
            Type: 'AWS::EC2::Instance'
            Properties:
              Size: real big
              Cidr: 0.2.2.2
            Tags:
              Name: tagname2
              Value: tagval2
        - Endpoint:
            Ref: https
          Protocol: https
        - Foo:
            Bar: bazval
            Dog: catval
            ParamTest2: undefined
        - Foo:
            Bar: mapbaz
            Dog: mapcat
            ParamTest2: undefined
  ParamTest: undefined
  ELBSecurityGroup:
    Type: 'AWS::EC2::SecurityGroup'
    Properties:
      GroupDescription: Allow access to the ELB
      VpcId:
        Ref: VPCID
      SecurityGroupIngress:
        - IpProtocol: tcp
          FromPort: '80'
          ToPort: '80'
          CidrIp: 0.0.0.0/0
        - IpProtocol: tcp
          FromPort: '443'
          ToPort: '443'
          CidrIp: 0.0.0.0/0
Outputs:
  ELBDNSName:
    Description: ELB DNS Name
    Value:
      'Fn::GetAtt':
        - ElasticLoadBalancer
        - DNSName
"
`;

exports[`should have a cli interface that works 1`] = `
"AWSTemplateFormatVersion: '2010-09-09'
dog:
  Type: 'AWS::EC2::Instance'
  Properties:
    Size: big
    Cidr: 0.1.1.1
  Tags:
    Name: tagname
    Value: tagval
cat:
  Type: 'AWS::EC2::Instance'
  Properties:
    Size: real big
    Cidr: 0.2.2.2
  Tags:
    Name: tagname2
    Value: tagval2
Resources:
  dog:
    Type: 'AWS::EC2::Instance'
    Properties:
      Size: big
      Cidr: 0.1.1.1
    Tags:
      Name: tagname
      Value: tagval
  cat:
    Type: 'AWS::EC2::Instance'
    Properties:
      Size: real big
      Cidr: 0.2.2.2
    Tags:
      Name: tagname2
      Value: tagval2
  NotificationTopic:
    Type: 'AWS::SNS::Topic'
    Properties:
      Subscription:
        - dog:
            Type: 'AWS::EC2::Instance'
            Properties:
              Size: big
              Cidr: 0.1.1.1
            Tags:
              Name: tagname
              Value: tagval
        - cat:
            Type: 'AWS::EC2::Instance'
            Properties:
              Size: real big
              Cidr: 0.2.2.2
            Tags:
              Name: tagname2
              Value: tagval2
        - Endpoint:
            Ref: https
          Protocol: https
        - Foo:
            Bar: bazval
            Dog: catval
            ParamTest2: myVal2
        - Foo:
            Bar: mapbaz
            Dog: mapcat
            ParamTest2: myVal2
  ParamTest: myVal
  ELBSecurityGroup:
    Type: 'AWS::EC2::SecurityGroup'
    Properties:
      GroupDescription: Allow access to the ELB
      VpcId:
        Ref: VPCID
      SecurityGroupIngress:
        - IpProtocol: tcp
          FromPort: '80'
          ToPort: '80'
          CidrIp: 0.0.0.0/0
        - IpProtocol: tcp
          FromPort: '443'
          ToPort: '443'
          CidrIp: 0.0.0.0/0
Outputs:
  ELBDNSName:
    Description: ELB DNS Name
    Value:
      'Fn::GetAtt':
        - ElasticLoadBalancer
        - DNSName
"
`;

exports[`should update var paths 1`] = `
"
  {{instances.0.name}}:
  Type: AWS::EC2::Instance
  Properties:
    Size: {{instances.0.size}}
    Cidr: {{instances.0.cidr}}
  Tags: {{instances.0, file:subTemplates/subSubTemplates/tags.yaml}}
  Foo: {{for:instances.0.tags, file:subTemplates/subSubTemplates/tags.yaml}}
  "
`;
